provisioner/data.js

```diff
@@ -1,30 +1,704 @@
+      var spotreqs = [];
+
+      res[0][region].Reservations.forEach(function(reservation) {
+        reservation.Instances.forEach(function(instance) {
+          instances.push(instance.InstanceId);
+          debug('Killing %s instance %s in %s',
+            that.workerType, instance.InstanceId, region);
+        });
+      });
+
+      res[1][region].SpotInstanceRequests.forEach(function(request) {
+        spotreqs.push(request.SpotInstanceRequestId);
+        debug('Cancelling %s spot request %s in %s',
+          that.workerType, request.SpotInstanceRequestId, region);
+      });
+
+      if (instances.length > 0) {
+        debug('Killing %d instances in %s', instances.length, region);
+        killinators.push(that.ec2.terminateInstances.inRegion(region, {
+          InstanceIds: instances,
+        }));
+      }
+
+      if (spotreqs.length > 0) {
+        debug('Cancelling %d spot requests in %s', spotreqs.length, region);
+        killinators.push(that.ec2.cancelSpotInstanceRequests.inRegion(region, {
+          SpotInstanceRequestIds: spotreqs,
+        }));
+      }
+
+    });
+    return Promise.all(killinators);
+  });
+
+  p = p.then(function(res) {
+    debug('Submitted kill and cancel requests for %s', that.workerType);
+  });
+
+  return p;
+
+};
+
+
+/**
+ * Provision a given workerType based on the pricing available, capacity existing
+ * and number of pending tasks.
+ */
+WorkerType.prototype.provision = function(debug, pricing, capacity, pending) {
+  assert(debug);
+  assert(pricing);
+  assert(typeof capacity === 'number');
+  assert(typeof pending === 'number');
+  var that = this;
+
+  var spotBids = this.determineSpotBids(debug, pricing, capacity, pending);
+
+  if (spotBids.length === 0) {
+    return [];
+  } else {
+    return Promise.all(spotBids.map(function(bid) {
+      return that.spawn(debug, bid);
+    }));
+  }
+};
+
+
+/**
+ * Create an AWS LaunchSpecification for this workerType.  This method
+ * does all the various overwriting of type and region specific LaunchSpecification
+ * keys.
+ */
+WorkerType.prototype.createLaunchSpec = function(debug, region, instanceType) {
+  assert(debug);
+  assert(region);
+  assert(instanceType);
+  return WorkerType.createLaunchSpec(debug, region, instanceType, this, this.keyPrefix);
+}
+
+
+/**
+ * Make sure that all combinations of LaunchSpecs work.  This sync
+ * function will throw if there is an error found or will return
+ * a dictionary of all the launch specs!
+ */
+WorkerType.prototype.testLaunchSpecs = function(debug) {
+  assert(debug);
+  return WorkerType.testLaunchSpecs(debug, this, this.keyPrefix);
+}
+
+
+/**
+ * We need to be able to create a launch specification for testing without
+ * already having an instance.  This is available as a non-instance method
+ * so that we can create and test launch specifications before inserting
+ * them into Azure.
+ */
+WorkerType.createLaunchSpec = function(debug, region, instanceType, worker, keyPrefix) {
```

Comment:
This is horribly ugly, but I love it... It looks to produce some really good and fairly readable error messages :)
(besides there is no way this can be pretty).

Depending on how or where you use this might want to returns errors as JSON instead.. At least that is more useful when API handers, but I suspect that is what the test one is for.